src/libraries/System.Runtime.InteropServices/gen/Common/ConvertToSourceGeneratedInteropFixer.cs

```diff
@@ -0,0 +1,263 @@
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.Interop.Analyzers
+{
+    public abstract class ConvertToSourceGeneratedInteropFixer : CodeFixProvider
+    {
+        public sealed override FixAllProvider GetFixAllProvider() => CustomFixAllProvider.Instance;
+
+        protected abstract string BaseEquivalenceKey { get; }
+
+        protected virtual IEnumerable<ConvertToSourceGeneratedInteropFix> CreateAllFixesForDiagnosticOptions(SyntaxNode node, ImmutableDictionary<string, Option> options)
+        {
+            // By default, we only have one fix for the options specified from the diagnostic.
+            yield return new ConvertToSourceGeneratedInteropFix(CreateFixForSelectedOptions(node, options), options);
+        }
+
+        protected abstract Func<DocumentEditor, CancellationToken, Task> CreateFixForSelectedOptions(SyntaxNode node, ImmutableDictionary<string, Option> selectedOptions);
+
+        protected abstract string GetDiagnosticTitle(ImmutableDictionary<string, Option> selectedOptions);
+
+        /// <summary>
+        /// A basic string-serializable option mechanism to enable the same fixer to be used for diagnostics with slightly different properties.
+        /// </summary>
+        public abstract record Option
+        {
+            private protected Option()
+            {
+            }
+
+            public const string AllowUnsafe = nameof(AllowUnsafe);
+            public const string MayRequireAdditionalWork = nameof(MayRequireAdditionalWork);
+
+            public sealed record Bool(bool Value) : Option
+            {
+                public override string ToString() => $"b:{Value};";
+            }
+
+            public sealed record String(string Value) : Option
+            {
+                public override string ToString() => $"s:{Value};";
+            }
+
+            public static ImmutableDictionary<string, Option> ParseOptionsFromEquivalenceKey(string equivalenceKey)
+            {
+                ImmutableDictionary<string, Option>.Builder options = ImmutableDictionary.CreateBuilder<string, Option>();
+                // The first ';' separates the base equivalence key from the options
+                foreach (string option in equivalenceKey.Split(';').Skip(0))
+                {
+                    string[] optionKeyAndValue = option.Split('=');
+                    if (optionKeyAndValue.Length != 2 || optionKeyAndValue[1].Length < 3)
+                    {
+                        continue;
+                    }
+
+                    string type = optionKeyAndValue[1].Substring(0, 2);
+                    string value = optionKeyAndValue[1].Substring(2);
+                    if (type == "b:")
+                    {
+                        options.Add(optionKeyAndValue[0], new Bool(bool.TryParse(value, out bool boolValue) && boolValue));
```

Comment:
If the parsing fails we just ignore this, right? I think we should just be correct and do `bool.Parse()` so we get an exception when things go wrong rather than being silent.

Reply:
I'll change this to `bool.Parse` since this is entirely within our own system (no user input here).